src/framework/Support/Collection.php

```diff
@@ -0,0 +1,368 @@
+    {
+        parent::__construct($array, static::ARRAY_AS_PROPS);
+    }
+
+    public static function create(array|object $array = [])
+    {
+        return collect($array);
+    }
+
+    /**
+     * Convert the collection to a string
+     * @method __toString
+     * @author Ben Hirst
+     *
+     * @return string
+     */
+    public function __toString(): string
+    {
+        return json_encode($this->getArrayCopy());
+    }
+
+    public function all()
+    {
+        return $this->toArray();
+    }
+
+    /**
+     *
+     * Creates a new Collection instance containing the same array
+     * @method copy
+     * @author Ben Hirst
+     *
+     * @return Skeletal\Support\Collection
+     */
+    public function copy(): static
+    {
+        return collect($this->getArrayCopy());
+    }
+
+    public function avg(string $key)
+    {
+        return $this->sum($key, static::ARRAY_SUM_AVG);
+    }
+
+    public function sum(string $key, int $flags = 0)
+    {
+        $items = $this->pluck($key)->filter()->flatten();
+        $sum = $items->array_sum();
+
+        if ($flags & static::ARRAY_SUM_AVG) {
+            return $sum/$items->count();
+        }
+
+        return $sum;
+    }
+
+    /**
+     * Loop through each item in the collection
+     */
+    public function each(callable $callback): void
+    {
+        foreach ($this->getIterator() as $key => $value) {
+            $callback($key, $value);
+        }
+    }
+
+    /**
+     * Loop through each item including nested arrays & objects in the collection
+     */
+    public function eachNested(callable $callback): void
+    {
+        foreach ($this->getIterator() as $key => $value) {
+            if (is_array($value)) {
+                $value = collect($value);
+            }
+
+            $callback($key, $value);
+
+            if ($value instanceof static) {
+                $value->eachNested($callback);
+            }
+        }
+    }
+
+    public function filter(?callable $callback = null)
+    {
+        return $this->array_filter($callback, ARRAY_FILTER_USE_BOTH);
+    }
+
+    public function flatten($depth = INF)
+    {
+        $result = [];
+
+        foreach ($this->toArray() as $item) {
+            $item = $item instanceof static ? $item->all() : $item;
+
+            if (! is_array($item)) {
+                $result[] = $item;
+            } else {
+                $values = $depth === 1
+                    ? array_values($item)
+                    : collect($item)->flatten($depth - 1);
+
+                foreach ($values as $value) {
+                    $result[] = $value;
+                }
+            }
+        }
+
+        return collect($result);
+    }
+
+    /**
+     * Creates a new Collection instance containing the keys of the array
+     */
+    public function keys(): static
+    {
+        return $this->array_keys();
+    }
+
+    public function map(callable $callback)
+    {
+        return array_map($callback, $this->getArrayCopy());
```

Comment:
Maybe add a flag for internal calls so that it can return an instance of Collection otherwise.. Usually calls to this method need an array returned, however the end user may prefer it to return a Collection like most other methods do